### DISCUSSION TITLE
defining whiich PIN to use on the PI

### DISCUSSION THREAD

**Discussion by neurodavidus** [wehr-lab/autopilot#83] [Q&A] [answered]
Hi Guys,

Looking in the blink task and others, I found it very difficult to understand how and where to define which PIN(s) on the PI board are going to be used to blink LED, control solenoid, etc ...

This is the part that seems to explain it but I don't get it:
![image](https://user-images.githubusercontent.com/4154936/109044109-50753380-76d2-11eb-9940-e580acce9054.png)

in the commented example pin 1 would correspond to what? 3V3 power or GPIO 1 ?  None of those pins are the first ones that come to my mind when wanting to blink a LED so I am bit confused there....

Describing in detail this step in a coupe of task would be really great. thanks

David

**Comment by sneakers-the-rat** [comment, marked as answer]
All numbering in autopilot is board numbering, which you're right does correspond to the 3.3V pin and maybe is a bad example, just supposed to illustrate "put some integer here" not literally pin 1. 

So the `Task.HARDWARE` class attribute is a generic description of what hardware is needed to do a task, but to adapt to different installations/setups/etc of pilots it doesn't specify how a piece of hardware is to be configured -- so the task may ask for an LED, but doesn't specify which pin it must be plugged into, etc. Instead the particular properties for a piece of hardware are defined in the `prefs.json` file in the autopilot user directory (typically `~/autopilot` ). both the `Task.HARDWARE` and `prefs.json` hardware descriptions have two levels, a "group" (like "LEDS" in this example) and an "ID" (like "dLED" in this example, the name of the particular hardware objected within that group) that have to match.

So, maybe going the other direction, you want to have some LED in your task, say you want to call it `"BLINKY"` . You plug `"BLINKY"` in to pin 7 on your raspberry pi and want to use it in a task. First you would tell autopilot about the hardware by adding it to the `HARDWARE` dictionary within `prefs.json` like (excluding other possible parameters for brevity's sake):

```json
"HARDWARE": {
    "LEDS": {
        "BLINKY": {
            "pin": 7
        }
    }
}
```

So to use it in your task you would declare it in the `HARDWARE` attribute of your class in a way that matches your `prefs.json` declaration like:

```python
HARDWARE = {
    'LEDS': {
        'BLINKY': gpio.PWM
    }
}
```

which will (literally) tell autopilot to make the hardware object available to you like this

```python
self.hardware['LEDS']['BLINKY'] = gpio.PWM(**prefs.get('HARDWARE')['LEDS']['BLINKY']))
```
